Convert rails 2 finders to arel for rails 3/4

jrafanie opened the pull request:
Note: this commit was required to be able to use arel_converter to mass convert some of the remaining rails 2 finders:  https://github.com/jrafanie/manageiq/commit/73968a0ca0ba5e09a8166281c2fb07ae492f634e

I can split this PR up if needed.

jrafanie commented on vmdb/app/models/miq_server/role_management.rb:
```diff
@@ -153,7 +153,7 @@ def role=(val)
   def assign_role(server_role, priority = nil)
     server_role          = ServerRole.to_role(server_role)
-    assigned_server_role = self.assigned_server_roles.find_or_create_by_server_role_id(server_role.id)
+    assigned_server_role = assigned_server_roles.find_or_create_by(:server_role_id => server_role.id)
```
@tenderlove find_or_create_by doesn't exist on rails 3.2.  I'm thinking we need to undo these changes on master and do them in rails4 branch, thoughts?

tenderlove replied:
Ya, you're right.  This will need to be on the rails4 branch.

jrafanie commented:
@Fryguy this is finally ready.. you might want to pass ?w=1 to ignore whitespace

cc @tenderlove

Fryguy commented on vmdb/app/models/bottleneck_event.rb:
```diff
@@ -6,7 +6,7 @@ class BottleneckEvent < ActiveRecord::Base
 
   def self.last_created_on(obj)
-    event = self.find(:first, :conditions => ["resource_type = ? AND resource_id = ?", obj.class.name, obj.id], :order => "created_on DESC")
+    event = self.where(["resource_type = ? AND resource_id = ?", obj.class.name, obj.id]).order("created_on DESC").first
```
Should we change the where clause to use Hash notation?

jrafanie replied:
Not now, too many moving pieces.  I started doing that and it's way too easy to get buried in shiny object changes.

Fryguy replied:
Ok very good...I'll review with those two things in mind.

Fryguy commented on vmdb/app/models/chargeback.rb:
```diff
@@ -235,7 +235,7 @@ def self.get_time_range(perf, interval, tz)
   def self.default_rates()
     rates = []
-    ChargebackRate.find(:all, :conditions=>{:default=>true}).each do |rate|
+    ChargebackRate.find(:all, :conditions => {:default => true}).each do |rate|
```
This one didn't get changed...

jrafanie replied:
yeah, this is probably the 2nd of many PRs to change to arel syntax.  The rubocop space around operator commit changed this one.

Fryguy replied:
Ah ok, then I'll ignore those for now.

jrafanie commented:
Note: I used https://github.com/jrafanie/arel_converter/tree/fixes which fixes/hacks some of the things that arel_converter wants to do like change :a => :b to :a: :b a few others.  arel_converter easily had 50+ instances of finders in just the models directory that it couldn't parse properly.  Even then, it totally messes up on .all and .find(options_hash).  It does handle many common conversions properly.

Fryguy commented on vmdb/app/models/cim_computer_system.rb:
```diff
@@ -283,7 +283,7 @@ def storage_managers
       next unless (/http:\/\/([^\/]*)\/na_admin/ =~ ai)
-      sma = NetappRemoteService.find(:all, :conditions => ["ipaddress = ? or hostname = ?", $1, $1])
+      sma = NetappRemoteService.where("ipaddress = ? or hostname = ?", $1, $1)
       storageManagers.concat(sma) unless sma.length == 0
```
Does concat work with an AR::Relation?

jrafanie replied:
Yes.

```ruby
irb(main):027:0> pf = MiqProductFeature.limit(2).select(:id)
=> [#<MiqProductFeature id: 1>, #<MiqProductFeature id: 2>]
irb(main):028:0> pf.class
=> ActiveRecord::Relation
irb(main):029:0> arr = []
=> []
irb(main):030:0> arr.concat(pf)
=> [#<MiqProductFeature id: 1>, #<MiqProductFeature id: 2>]
irb(main):031:0> arr.first.class
=> MiqProductFeature(id: integer, ...)
```

Fryguy commented on vmdb/app/models/classification.rb:
```diff
@@ -176,17 +176,18 @@ def self.create_category!(options)
-    self.in_region(region_id).all(find_opts).collect { |c| c.name if c.tag2ns(c.tag.name) == ns }.compact
+    self.in_region(region_id).
+      where(:parent_id => 0, :perf_by_tag => true).
+      includes(:tag).collect { |c| c.name if c.tag2ns(c.tag.name) == ns }.compact
```
Should each dotted method call be on a separate line (also the guide says put the dot on the next line)...

```ruby
    self.in_region(region_id)
      .where(:parent_id => 0, :perf_by_tag => true)
      .includes(:tag)
      .collect { |c| c.name if c.tag2ns(c.tag.name) == ns }
      .compact
```

jrafanie replied:
Ah, I didn't do it... but I'll fix it.  Must be that Seattle coffee.  cough @tenderlove

Fryguy commented on vmdb/app/models/condition.rb:
```diff
@@ -29,7 +29,7 @@ def applies_to?(rec, inputs={})
 
   def self.conditions
-    find(:all).collect {|c| c.expression}
+    pluck(:expression)
```
I think expression is a serialized column.  Does pluck do the serialization or does it do raw SQL value.  Trying to keep that in mind when I review pluck changes.  I'm not sure they call the getters if they are overridden.

jrafanie replied:
Yes, it looks like pluck handles it for us.  See below...

Note, this `["expression", "--- !ruby/string:VimString\nstr: b\nxsiType: \nvimType: \n"]` in the insert statement below:

```ruby
irb(main):008:0> Condition.create!(:description => "aaa", :modifier => "deny", :expression => VimString.new("b"), :towhat => "b")
...
  SQL (0.3ms)  INSERT INTO "conditions" ("applies_to_exp", "created_on", "description", "expression", "file_mtime", "filename", "guid", "miq_policy_id", "modifier", "name", "notes", "towhat", "updated_on") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13) RETURNING "id"  [["applies_to_exp", nil], ["created_on", Thu, 15 Jan 2015 20:05:07 UTC +00:00], ["description", "aaa"], ["expression", "--- !ruby/string:VimString\nstr: b\nxsiType: \nvimType: \n"], ["file_mtime", nil], ["filename", nil], ["guid", "cd7614ec-9cf1-11e4-9d2e-3c15c2c9babc"], ["miq_policy_id", nil], ["modifier", "deny"], ["name", "cd7614ec-9cf1-11e4-9d2e-3c15c2c9babc"], ["notes", nil], ["towhat", "b"], ["updated_on", Thu, 15 Jan 2015 20:05:07 UTC +00:00]]
   (0.8ms)  COMMIT
=> #<Condition id: 4, name: "cd7614ec-9cf1-11e4-9d2e-3c15c2c9babc", description: "aaa", modifier: "deny", expression: "b", towhat: "b", file_mtime: nil, created_on: "2015-01-15 20:05:07", updated_on: "2015-01-15 20:05:07", guid: "cd7614ec-9cf1-11e4-9d2e-3c15c2c9babc", filename: nil, applies_to_exp: nil, miq_policy_id: nil, notes: nil>
irb(main):009:0> Condition.conditions
  SQL (0.3ms)  SELECT "conditions"."expression" FROM "conditions" WHERE ("conditions".id >= 0 AND "conditions".id <= 999999999999)
=> ["b"]
irb(main):010:0> Condition.conditions.first.class
  SQL (0.4ms)  SELECT "conditions"."expression" FROM "conditions" WHERE ("conditions".id >= 0 AND "conditions".id <= 999999999999)
=> VimString
```

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory.rb:
```diff
@@ -176,7 +176,7 @@ def save_guest_devices_inventory(hardware, hashes)
     end
 
-    deletes = hardware.guest_devices.find(:all, :conditions => {:device_type => ['ethernet', 'storage']}).dup
+    deletes = hardware.guest_devices.where(:device_type => ["ethernet", "storage"]).to_a.dup
```
Is the dup even necessary anymore?  I thought to_a returned a new Array, or does it return a reference to the cached Array inside of the AR::Relation.

jrafanie replied:
It appears to return the same array for the same AR::Relation, so, maybe the dup is still needed, cc @tenderlove @matthewd 

```ruby
irb(main):023:0> pf = MiqProductFeature.limit(2).select(:id)
=> [#<MiqProductFeature id: 1>, #<MiqProductFeature id: 2>]
irb(main):024:0> pf.to_a.object_id
=> 70215016980300
irb(main):025:0> pf.to_a.object_id
=> 70215016980300
```

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory_infra.rb:
```diff
@@ -77,8 +77,8 @@ def save_storages_inventory(ems, hashes, target = nil)
-    locs  = Storage.all(:conditions => ["location IN (?)", locs]) unless locs.empty?
-    names = Storage.all(:conditions => ["location IS NULL AND name IN (?)", names]) unless names.empty?
+    locs  = Storage.where("location IN (?)", locs) unless locs.empty?
+    names = Storage.where("location IS NULL AND name IN (?)", names) unless names.empty?
```
I think this one is important to change to hash notation in the followup PR.

Fryguy commented on vmdb/app/models/ext_management_system.rb:
```diff
@@ -225,7 +225,7 @@ def clustered_hosts
 
   def miq_proxies
-    MiqProxy.find(:all).collect {|p| p if p.ext_management_system == self}.compact
+    MiqProxy.all.select { |p| p.ext_management_system == self }
```
This is screaming for an extra `.includes(:ext_management_system)`  I know it's shiny, but let's change in the follow up PR.

Fryguy commented on vmdb/app/models/ext_management_system.rb:
```diff
@@ -281,7 +281,7 @@ def resource_pools_non_default
       self.resource_pools.select { |r| !r.is_default }
     else
-      self.resource_pools.all(:conditions => ["is_default != ?", true])
+      self.resource_pools.where("is_default != ?", true).to_a
```
I think this is actually a bug in Arel or maybe AR::Relation (that is, querying a boolean column with a literal true or false).  I was trying to write a query for this the other day in a similar fashion and it doesn't return what is expected.  I had meant to put together a bug report, and I totally forgot.

jrafanie replied:
¯\_(ツ)_/¯

Fryguy commented on vmdb/app/models/host.rb:
```diff
@@ -761,7 +761,7 @@ def is_proxy_active?
 
   def miq_proxies
-    MiqProxy.find(:all).collect {|p| p if p.hosts.include?(self)}.compact
+    MiqProxy.all.select { |p| p.hosts.include?(self) }
```
Also screams for `.includes(:hosts)`

Fryguy commented on vmdb/app/models/host_microsoft.rb:
```diff
@@ -13,10 +13,10 @@ def verify_credentials(auth_type=nil, options={})
   def verify_credentials_task(*authentication)
-    svr_list = MiqServer.find(:all).collect {|s| s.has_role?(authentication_check_role) ? s : nil}.compact
-    raise "No #{Dictionary::gettext("miq_servers", :type=>:table)} were found to verify Windows credentials." if svr_list.blank?
+    svr_list = MiqServer.all.select { |s| s.has_role?(authentication_check_role) }
```
More .includes here (include whatever is needed for roles stuff)

Fryguy commented on vmdb/app/models/job.rb:
```diff
@@ -55,7 +55,7 @@ def check_active_on_destroy
   def self.agent_state_update_queue(jobid, state, message=nil)
     begin
-      job = Job.find(:first, :conditions => ["guid = ?", jobid], :select => "id, state, guid")
+      job = Job.where("guid = ?", jobid).select("id, state, guid").first
```
Need an Array for `.select`.

ASIDE: @tenderlove Is there a reason that AR::Relation#select requires an Array and not just splatted args?  .pluck, .order, and others take splatted args, and it's confusing to have select act differently.  It's so easy to make a mistake.

jrafanie replied:
No, there is only one situation that doesn't work:  select with multiple arguments (the referenced code above has a single string with comma separated values):

```ruby

irb(main):006:0> MiqProductFeature.select("id, name").take(2)
=> [#<MiqProductFeature id: 1, name: "Everything">, #<MiqProductFeature id: 2, name: "Workloads">]

irb(main):008:0> MiqProductFeature.select(["id", "name"]).take(2)
=> [#<MiqProductFeature id: 1, name: "Everything">, #<MiqProductFeature id: 2, name: "Workloads">]

irb(main):009:0> MiqProductFeature.select([:id, :name]).take(2)
=> [#<MiqProductFeature id: 1, name: "Everything">, #<MiqProductFeature id: 2, name: "Workloads">]

```

This fails:
```ruby
irb(main):010:0> MiqProductFeature.select(:id, :name).take(2)
ArgumentError: wrong number of arguments (2 for 0..1)

irb(main):011:0> MiqProductFeature.select('id', 'name').take(2)
ArgumentError: wrong number of arguments (2 for 0..1)
```

The confusing part is that I believe the other arel methods like order, includes, etc. accept multiple arguments (this latter failing case).

jrafanie replied:
Note, pluck, is just like select above.

Fryguy commented on vmdb/app/models/metric/capture.rb:
```diff
@@ -37,7 +37,7 @@ def self.capture_threshold(target)
     log_header = "MIQ(#{self.name}.perf_capture_health_check)"
     zone ||= MiqServer.my_server.zone(true)
-    q_items = MiqQueue.all(:select => "created_on, args", :conditions => {:state => "ready", :role => "ems_metrics_collector", :method_name => "perf_capture", :zone => zone.name}, :order => "created_on ASC")
+    q_items = MiqQueue.select("created_on, args").where(:state => "ready", :role => "ems_metrics_collector", :method_name => "perf_capture", :zone => zone.name).order("created_on ASC")
```
Need an Array for `.select`

jrafanie replied:
Same as above, it's a single string with CSVs

Fryguy replied:
Weird...I saw it as two separate params...not sure why I commented on this one. :eyes_glazing_over:

Fryguy commented on vmdb/app/models/metric/ci_mixin/state_finders.rb:
```diff
@@ -26,7 +26,7 @@ def vim_performance_state_for_ts(ts)
 
   def preload_vim_performance_state_for_ts(conditions = {})
-    @states_by_ts = self.vim_performance_states.all(:conditions => conditions).index_by(&:timestamp)
+    @states_by_ts = self.vim_performance_states.where(conditions).index_by(&:timestamp)
```
If conditions is an empty Hash as per the default for the param, will that cause a problem?

jrafanie replied:
Appears to be ok, unless it's different for an association.

```ruby
irb(main):016:0> MiqProductFeature.all(:conditions => {}).count
=> 712
irb(main):017:0> MiqProductFeature.where({}).count
=> 712
```

Fryguy commented on vmdb/app/models/metric/rollup.rb:
```diff
@@ -312,7 +312,7 @@ def self.find_distinct_resources
 
     recs.each_with_object([]) do |(klass, ids), ret|
-      ret.concat(klass.constantize.find_all_by_id(ids))
+      ret.concat(klass.constantize.where(:id => ids).all)
```
Is the .all necessary here?

jrafanie replied:
No, i'll fix

Fryguy commented on vmdb/app/models/miq_alert.rb:
```diff
@@ -152,7 +152,7 @@ def self.evaluate_hourly_timer
         model = a.db.constantize
         unless model.name == "Storage" # No clean way to get zone from storage
-          targets += model.all(:include => [:ext_management_system]).select {|i| i.my_zone == zone}
+          targets += model.includes(:ext_management_system).select { |i| (i.my_zone == zone) }
```
Unnecessary parens,

jrafanie replied:
ok, fixing

Fryguy commented on vmdb/app/models/miq_bulk_import.rb:
```diff
@@ -52,7 +52,7 @@ def self.upload(fd, tags, keys)
     primary_key, primary_key_value = keys2array.shift
-    result = klass.all(:conditions => "LOWER(#{primary_key}) LIKE '#{primary_key_value.downcase}'")
+    result = klass.where("LOWER(#{primary_key}) LIKE '#{primary_key_value.downcase}'")
     return result if result.empty? || result.length == 1
```
Will this cause two queries or does the `.empty?` deal with caching?  Perhaps the code should just do `.size <= 1` to be safe.

Fryguy commented on vmdb/app/models/miq_event.rb:
```diff
@@ -192,8 +192,8 @@ def self.event_name_for_target(target, event_suffix)
-    p_ids = MiqPolicyContent.find_all_by_miq_event_id(self.id).collect {|pc| pc.miq_policy_id}.uniq
-    MiqPolicy.find(:all, :conditions => {:id => p_ids})
+    p_ids = MiqPolicyContent.where(:miq_event_id => self.id).uniq.pluck(:miq_policy_id)
+    MiqPolicy.where(:id => p_ids).to_a
```
I think MiqPolicy objects are gettable from their content, so the two methods could be made into 1...(maybe separate PR)

```ruby
MiqPolicyContent.where(:miq_event_id => self.id).includes(:miq_policy).collect(&:miq_policy).uniq
```

ASIDE: Can you pluck relationships or is it strictly columns?

jrafanie replied:
cc @tenderlove, I don't know

Fryguy commented on vmdb/app/models/miq_expression.rb:
```diff
@@ -1652,7 +1652,10 @@ def self.get_entry_details(field)
-      return Tag.find(:all, :conditions => ["name like ?", "/user/#{cat}%"]).collect {|t| [t.name.split("/").last, t.name.split("/").last]}
+      return Tag.where("name like ?", "/user/#{cat}%").select(:name).collect do |t|
+        tag_name = t.name.split("/").last
+        [tag_name, tag_name]
```
o_O

jrafanie replied:
It hurt my eyes to leave it like it was

Fryguy commented on vmdb/app/models/miq_group.rb:
```diff
@@ -64,9 +64,9 @@ def self.allows?(group, options={})
     else
-      groups = group.resource.miq_groups.find(:all, :order => "sequence DESC")
+      groups = group.resource.miq_groups.order("sequence DESC")
     end
```
You don't even need all the groups, but this works just fine...thinking it could be cleaned up as:

```ruby
    groups =
      if group.resource.nil?
        where(:resource_id => nil, :resource_type => nil)
      else
        group.resource.miq_groups
      end
    last_group = groups.order("sequence DESC").first
    group.sequence = last_group.nil? ? 1 : last_group.sequence + 1
```

Also, seeing a possible race condition here... cc @gmcculloug, @bzwei

Fryguy commented on vmdb/app/models/miq_host_provision_workflow.rb:
```diff
@@ -75,12 +75,15 @@ def allowed_hosts(options={})
 
   def allowed_ws_hosts(options={})
-    Host.find(:all, :conditions=>["mac_address is not NULL"]).find_all {|h| h.ipmi_enabled}
+    Host.where("mac_address is not NULL").select { |h| h.ipmi_enabled }
```
Could this be query-ized to avoid pulling everything back?

Something like `Host.where("mac_address is not NULL").where(:ipmi_enabled => true)`

Although this might fall into the Arel bug I was talking about earlier.

jrafanie replied:
Note, ipmi_enabled is an alias of ipmi_config_valid? which is a method with logic basic on the ipmi_address column, ipmi_enabled is not a column...

Fryguy commented on vmdb/app/models/miq_host_provision_workflow.rb:
```diff
@@ -75,12 +75,15 @@ def allowed_hosts(options={})
+
+    ExtManagementSystem.select("id, name").each do |e|
+      result[e.id] = e.name if e.kind_of?(EmsVmware)
+    end
```
This pattern is so common I feel like Ruby should just have a method for.  It's like `index_by` except that the value is a property and not the whole object.  poke poke @tenderlove :wink:

Fryguy replied:
Also, the if condition can be removed by saying `EmsVmware.select("id, name")...`

jrafanie replied:
Shiny object!  :sparkles:

Fryguy commented on vmdb/app/models/miq_provision_workflow.rb:
```diff
@@ -776,7 +776,7 @@ def allowed_snapshots(options={})
   def allowed_datastore_storage_controller(options={})
     result = {}
-    NetAppFiler.find(:all).each {|f| result[f.name] = f.name}
+    NetAppFiler.select(:name).each { |f| result[f.name] = f.name }
```
Can just use .pluck, right?  `NetAppFiler.pluck(:name).index_by { |n| n }`

jrafanie replied:
yeah, let's fix these :sparkles: in a followup pr.

Fryguy commented on vmdb/app/models/miq_provision_workflow.rb:
```diff
@@ -1260,7 +1260,7 @@ def ws_template_fields(values, fields, ws_values)
       data[:placement_cluster_name] = placement_cluster_name.to_s.downcase
       $log.info "#{log_header} placement_cluster_name:<#{data[:placement_cluster_name].inspect}>"
-      data[:data_centers] = EmsCluster.find(:all, :conditions => ["lower(name) = ?", data[:placement_cluster_name]]).collect {|c| c.v_parent_datacenter}
+      data[:data_centers] = EmsCluster.where("lower(name) = ?", data[:placement_cluster_name]).collect { |c| c.v_parent_datacenter }
```
Could use `.includes(:v_parent_datacenter)`, then at the end you can do `.collect(&:v_parent_datacenter)`

jrafanie replied:
:sparkles:

Fryguy commented on vmdb/app/models/miq_queue.rb:
```diff
@@ -220,10 +220,10 @@ def self.peek(options = {})
-    MiqQueue.all(args)
+    result = MiqQueue.where(cond).order(:priority, :id)
+    result.select(select) unless select.nil?
+    result.limit(limit || 1)
```
The limit can be attached to the first line instead of a separate line.

Fryguy commented on vmdb/app/models/miq_region.rb:
```diff
@@ -122,7 +122,7 @@ def role_assigned?(role_name)
 
   def remote_ui_miq_server
-    MiqServer.in_region(self.region).first(:conditions => {:has_active_userinterface => true})
+    MiqServer.in_region(self.region).where(:has_active_userinterface => true).first
```
We should double check that `.where(x => true)` bug I've mentioned a few times.

Fryguy commented on vmdb/app/models/miq_report/generator.rb:
```diff
@@ -670,7 +670,10 @@ def build_add_includes(data_records, entry, includes)
+        @descriptions_by_tag_id ||= Classification.where("parent_id != 0").inject({}) do |h, c|
+          h[c.tag_id] = c.description
+          h
+        end
```
So many shiny objects (each_with_object) :sparkles:

Fryguy commented on vmdb/app/models/miq_server/role_management.rb:
```diff
@@ -67,9 +67,9 @@ def set_role_activation(active, *roles)
 
-    roles = (roles.length == 1 && roles[0] == "*") ? self.server_roles : ServerRole.find_all_by_name(roles)
+    roles = (roles.length == 1 && roles[0] == "*") ? self.server_roles : ServerRole.where(:name => roles)
     ids   = roles.collect { |r| r.id }
```
Is the `roles` variable used elsewhere aside from the .collect here?  If not, should the query(ies) have `.select(:id)` or `.pluck(:id)`?

Fryguy commented on vmdb/app/models/miq_server/role_management.rb:
```diff
@@ -67,9 +67,9 @@ def set_role_activation(active, *roles)
+    roles = (roles.length == 1 && roles[0] == "*") ? self.server_roles : ServerRole.where(:name => roles)
     ids   = roles.collect { |r| r.id }
-    self.assigned_server_roles.find_all_by_server_role_id(ids).each do |a|
+    self.assigned_server_roles.where(:server_role_id => (ids)).each do |a|
```
Unnecessary params.

Fryguy commented on vmdb/app/models/miq_server/role_management.rb:
```diff
@@ -127,7 +127,7 @@ def role=(val)
       val = val.to_s.strip.downcase
       if val.blank?
-        self.server_roles.delete(ServerRole.find(:all))
+        self.server_roles.delete(ServerRole.all)
```
Can this just be `self.service_roles.delete_all`?

jrafanie replied:
I guess... how old is this code?  When was delete_all added?

Fryguy commented on vmdb/app/models/mixins/assignment_mixin.rb:
```diff
@@ -85,11 +85,11 @@ def namespace
       # Get all assigned, enabled instances for type klass
-      ass_hash = self.all(find_options).inject({}) {|h,a| h[a.id] = a; h}
-      tags = Tag.all(:conditions => ["taggings.taggable_type = ? and tags.name like ?", self.name, "#{self.namespace}/%"], :include => "taggings")
+      assignment_map = self.all(find_options).inject({}) { |h, a| h[a.id] = a; h }
```
- Should this be `all(find_options)` or `where(find_options)`
- The `.inject...` simplifies to `.index_by(&:id)`

jrafanie replied:
No, we need to create a helper library to pull out keys from the options hash to apply: .order, .includes, .select, etc.  I think @tenderlove did it in one place in this PR.   For now, I avoided changing all/first/find with an options hash until we have that library to extract and apply the sql correctly.

Fryguy commented on vmdb/app/models/mixins/event_mixin.rb:
```diff
@@ -35,7 +35,7 @@ def find_one_event(assoc, order)
     return nil if klass.blank?
 
-    event = klass.find(:first, :conditions => ewc, :order => order)
+    event = klass.where(ewc).order(order).first
```
Useless assignment to `event`

Fryguy commented on vmdb/app/models/registry_item.rb:
```diff
@@ -14,7 +14,7 @@ def self.add_elements(miq_set, scan_item, vm, xmlNode)
 
     new_reg = []
-    deletes = vm.registry_items.find(:all, :select=>"id, name").collect {|rec| [rec.id, rec.name]}
+    deletes = vm.registry_items.pluck(:id, :name)
```
Will pluck go back to the database if `registry_items` is cached?  If so, perhaps the collect can stay, and just remove the finder in the middle.

jrafanie replied:
:confused: I don't know

jrafanie replied:
cc @tenderlove @matthewd

matthewd replied:
I think it will re-query. But I'm not sure whether there's a good reason for that.

Fryguy commented on vmdb/app/models/repository.rb:
```diff
@@ -59,17 +59,17 @@ def path=(newpath)
   def vms
     if relative_path != "/"
-      Vm.all(:conditions=>["storage_id=? and location like ?", storage_id, relative_path+"%"])
+      Vm.where("storage_id=? and location like ?", storage_id, (relative_path + "%")).to_a
```
Useless parens (unless you think it makes it clearer, but in that case we should just interpolate.)

Fryguy commented on vmdb/app/models/repository.rb:
```diff
@@ -59,17 +59,17 @@ def path=(newpath)
   def miq_templates
     if relative_path != "/"
-      MiqTemplate.all(:conditions=>["storage_id=? and location like ?", storage_id, relative_path+"%"])
+      MiqTemplate.where("storage_id=? and location like ?", storage_id, (relative_path + "%")).to_a
```
Same on the parens.

Fryguy commented on vmdb/app/models/rss_feed.rb:
```diff
@@ -82,7 +82,7 @@ def self.roles
     #   roles.push(t.name.split("/").last)
     # }
-    Tag.find(:all, :conditions=>"name like '/managed/roles/%'").each {|t| roles.push(t.name.split("/").last)}
+    Tag.where("name like '/managed/roles/%'").each { |t| roles.push(t.name.split("/").last) }
```
Is all this just a collect on a pluck?  `Tag.where("name...").pluck(:name).collect { |n| n.split("/").last }`

Fryguy commented on vmdb/app/models/server_role.rb:
```diff
@@ -24,7 +24,7 @@ def self.seed
         cols.each_index {|i| action[cols[i].to_sym] = arr[i]}
 
-        rec = self.find_by_name(action[:name])
+        rec = self.where(:name => (action[:name])).first
```
Useless parens

Fryguy commented on vmdb/app/models/server_role.rb:
```diff
@@ -52,23 +52,23 @@ def self.to_role(server_role)
 
   def self.all_names
-    self.find(:all, :order => :name, :select => "id, name").collect { |r| r.name }
+    self.order(:name).pluck(:name)
```
I remember a case where we had to .select the "id" in the past even if we didn't use it because one of the callbacks (like after_find) needed it.  I can't remember if it was this class or not, but something to keep in mind to check when we remove select values.

jrafanie replied:
Good point, it's probably better to put it back when it fails then to proactively put possibly unused selects.

Fryguy commented on vmdb/app/models/server_role.rb:
```diff
@@ -52,23 +52,23 @@ def self.to_role(server_role)
 
   def self.database_scoped_roles
-    @database_scoped_roles ||= self.find_all_by_role_scope('database').sort { |a,b| a.name <=> b.name }
+    @database_scoped_roles ||= self.where(:role_scope => ('database').order(:name)).to_a
```
I think the parens are wrong here.  The parens on database are useless and when you remove them you see .order being called on a String.

Fryguy commented on vmdb/app/models/service/retirement_management.rb:
```diff
@@ -4,7 +4,7 @@ module Service::RetirementManagement
   module ClassMethods
     def retirement_check
-      services = Service.all(:conditions => ["(retires_on IS NOT NULL OR retired = ?)", true])
+      services = Service.where("(retires_on IS NOT NULL OR retired = ?)", true)
```
useless parens in the query string.  Also, the next line can just be `services.each(&:retirement_check)` :sparkles:

Fryguy commented on vmdb/app/models/snia_file_share.rb:
```diff
@@ -260,7 +260,7 @@ def storage_count
 
   def applicable_hosts
-    Host.all(:select   => "name, id").inject({}) {|h, host| h[host.id] = host.name; h}
+    Host.select("name, id").inject({}) { |h, host| h[host.id] = host.name; h }
```
So many shiny objects it's killing me :dizzy:

Fryguy commented on vmdb/app/models/storage.rb:
```diff
@@ -67,7 +67,7 @@ class Storage < ActiveRecord::Base
 
   def miq_proxies
-    MiqProxy.find(:all).select { |p| p.storages.include?(self) }
+    MiqProxy.all.select { |p| p.storages.include?(self) }
```
This should have a `.includes(:storages)` (if storages are a real or virtual relationship here)

Fryguy commented on vmdb/app/models/storage.rb:
```diff
@@ -360,24 +360,24 @@ def self.scan_timer(zone_name = nil)
       self.storage_files.select { |f| f.ext_name == "vmx" && f.vm_or_template_id.nil? }
     else
-      self.storage_files.all(:conditions => {:ext_name => "vmx", :vm_or_template_id => nil})
+      self.storage_files.where(:ext_name => "vmx", :vm_or_template_id => nil)
```
Since only the name is used after, maybe this should have a `.select(:name)`

Fryguy replied:
Then again maybe not...I think the idea was to prevent AR from going back to the database all the time, and calling storage_files caches it once, so the other methods (like `managed_vm_config_files`) can just do an in-memory select on the cache.

Fryguy commented on vmdb/app/models/vim_performance_daily.rb:
```diff
@@ -59,8 +59,24 @@ def self.find_by_time_profile(*args)
+      scope.all
+    else
+      raise "Unsupported finder value #{cnt}"
+    end
```
Can we make a generic method that is shared by every activerecord subclass?  Or a module that does this kind of like @tenderlove did MiqPreloader?  I know you have the FIXME there, but it doesn't seem much of a stretch to extract it now.  I know @tenderlove had written that apply_options method elsewhere (can't remember where), and that has very similar logic.

Additionally, we should be careful with deleting from the options.  The callers may not be expecting that.  Not sure if options is passed in or not, so that may not apply here, but if we move to a module or mixin then we would have to be aware of that.  We can probably use a method on the hash or the keys like .except, .slice, .include_any? or something like that to simplify the validation without deleting it.

Fryguy replied:
If it's pushed up a a mixin to ActiveRecord::Base a la the ar_* extensions, then this code becomes something like `klass.apply_options(options).to_a`

Fryguy commented on vmdb/app/models/vim_performance_daily.rb:
```diff
@@ -59,8 +59,24 @@ def self.find_by_time_profile(*args)
+      scope = scope.select(select) if select
+      scope = scope.order(order) if order
+      scope = scope.where(wheres) if wheres
+      scope.all
```
Should this return `.to_a`?

Fryguy commented on vmdb/app/models/vm.rb:
```diff
@@ -48,7 +48,7 @@ def self.find_all_by_mac_address_and_hostname_and_ipaddress(mac_address, hostnam
     conds[0] = "(#{conds[0].join(" AND ")})"
 
-    Hardware.all(:include => include.uniq, :conditions => conds).collect { |h| h.vm_or_template.kind_of?(Vm) ? h.vm_or_template : nil }.compact
+    Hardware.includes(include.uniq).where(conds).collect { |h|  h.vm_or_template.kind_of?(Vm) ? h.vm_or_template : nil}.compact
```
That include variable is making me go o_O.  We should probably rename that guy.

Fryguy commented on vmdb/app/models/vm_or_template.rb:
```diff
@@ -1828,7 +1828,7 @@ def perf_rollup_parent(interval_name=nil)
   def self.vms_by_ipaddress(ipaddress)
     ipaddresses = ipaddress.split(',')
-    Network.find(:all, :conditions => ["ipaddress in (?)", ipaddresses]).each do |network|
+    Network.where(["ipaddress in (?)", ipaddresses]).each do |network|
```
Are the `[]` needed?

Fryguy commented on vmdb/app/models/vm_or_template/retirement_management.rb:
```diff
@@ -6,7 +6,7 @@ module ClassMethods
       zone    = MiqServer.my_server.zone
       ems_ids = zone.ext_management_systems(:select => :id).collect { |e| e.id }.flatten
-      vms     = Vm.all(:conditions => ["(retires_on IS NOT NULL OR retired = ?) AND ems_id IN (?)", true, ems_ids])
+      vms     = Vm.where(["(retires_on IS NOT NULL OR retired = ?) AND ems_id IN (?)", true, ems_ids])
```
Same question on the `[]`

Fryguy commented on vmdb/app/models/zone.rb:
```diff
@@ -202,7 +202,7 @@ def storages
   def miq_proxies
     ems_ids = self.ext_management_systems.collect {|e| e.id }
-    MiqProxy.all(:include => :host, :conditions => ["hosts.ems_id in (?)", ems_ids])
+    MiqProxy.includes(:host).where("hosts.ems_id in (?)", ems_ids).to_a
```
Should this be `.includes`, `.joins` or `.references` (is that last one even a thing? I can't remember anymore, but I remember the long talks in rails about it).  cc @tenderlove

Fryguy commented on vmdb/db/migrate/20100908214807_recreate_relationships.rb:
```diff
@@ -72,7 +72,7 @@ def self.down
 
   def self.collect_relationships
-    relats = RelationshipsOrig.all(:include => [:parent, :child])
+    relats = RelationshipsOrig.includes([:parent, :child]).all
```
Useless `[]`

Fryguy commented on vmdb/db/migrate/20101028195038_remove_custom_fields_from_vms_hosts.rb:
```diff
@@ -38,7 +38,7 @@ def self.down
     say_with_time("Migrate #{klass.name} custom columns to CustomAttribute") do
       condition = (custom_names).join(' is not NULL OR ') +  ' is not NULL'
-      items = klass.find(:all, :select=>'id,' + custom_names.join(','), :conditions => condition)
+      items = klass.select('id,' + custom_names.join(',')).where(condition)
```
Instead of building a string, can we just use the array? `klass.select(custom_names + ["id"]).where(condition)`

Fryguy commented on vmdb/db/migrate/20110602123937_change_alert_processor_to_notifier.rb:
```diff
@@ -8,21 +8,21 @@ class ServerRole < ActiveRecord::Base; end
   def self.up
     say_with_time("Update MiqQueue messages with 'alert_processor' role to 'notifier' role") do
-      MiqQueue.update_all({:role => "notifier"}, {:role => "alert_processor"})
+      MiqQueue.where(:role => "alert_processor").update_all({:role => "notifier"})
```
Unnecessary `{}` here and on the next 3

Fryguy commented:
This is amazing @jrafanie :heart_eyes: .

Fryguy commented:
If you search the diff for `.where([` you can find a bunch of deletes for Useless `[]` on the where.

jrafanie commented:
For reference, this is the superfluous [] example:

```ruby
irb(main):022:0> MiqProductFeature.where(["name = ?", "Delete"]).pluck(:name).take(2)
=> ["Delete", "Delete"]
```

The [] is not needed:

```ruby
irb(main):023:0> MiqProductFeature.where("name = ?", "Delete").pluck(:name).take(2)
=> ["Delete", "Delete"]
```

jrafanie commented:
@Fryguy Finally went through everything.  Please review the last 7 commits starting with:

* Remove superfluous [] .where([]) -> .where()

In general, I fixed the smaller :sparkles: :lipstick:  changes but left off the missing .includes from the original code. I also rewrote history on the paren bug in the ServerRole.database_scoped_roles method.

Let me know if there's anything else that needs to be solved in this PR.

matthewd commented on vmdb/app/models/miq_queue.rb:
```diff
@@ -220,10 +220,9 @@ def self.peek(options = {})
-    args[:limit]  = limit || 1
-    MiqQueue.all(args)
+    result = MiqQueue.where(cond).order(:priority, :id).limit(limit || 1)
+    result.select(select) unless select.nil?
```
... that's not gonna do anything.

jrafanie replied:
:confused: what do you mean @matthewd?

Fryguy replied:
You need to set it back to result else it's a noop.   `result = result.select...`

jrafanie replied:
Thanks, my eyes need to look at something else.  I didn't see it.  :open_mouth:

Fryguy commented:
@jrafanie Can you pluck mutiple columns in Rails 3.2?  I can't locally...

```ruby
[1] pry(main)> Vm.limit(5).pluck([:id, :ems_ref])
  SQL (0.5ms)  SELECT id, ems_ref FROM "vms" WHERE "vms"."template" = 'f' AND "vms"."type" IN ('Vm', 'VmInfra', 'VmKvm', 'VmMicrosoft', 'VmRedhat', 'VmVmware', 'VmXen', 'VmCloud', 'VmAmazon', 'VmOpenstack') LIMIT 5
=> ["vm-101616", "vm-1023", "vm-1024", "vm-112803", "vm-112805"]
```

jrafanie commented:
Fine, I'll back out all the plucks with multiple columns.

http://meltingice.net/2013/06/11/pluck-multiple-columns-rails/

jrafanie commented:
Ok, @Fryguy @matthewd @tenderlove... anything else???

Fryguy commented:
:fire: in the hole!

gmcculloug commented:
@jrafanie This PR modifies calls to models backed by ActsAsArModel that do not support the newer calls.  Looks like we have 9 models that use that class.

For example:
```
[----] F, [2015-01-19T09:06:09.354127 #21209:3ffb85865be4] FATAL -- : Error caught: [NoMethodError] undefined method `pluck' for NetAppFiler:Class
/Users/gmccullough/work/manageiq/vmdb/app/models/miq_provision_workflow.rb:778:in `allowed_datastore_storage_controller'
```

jrafanie commented:
Thanks @gmcculloug, good times.  I"m thinking these need to be converted to ActiveModel since ActsAsArModel was our version of ActiveModel before there was an ActiveModel.

jrafanie commented:
Short-time, we can see if we .select(:column).collect { |c| c.column} fixes the issue

jrafanie commented:
cc @tenderlove

gmcculloug commented:
@jrafanie Tested the select option in irb and that fails:
```
irb(main):001:0> NetAppFiler.select(:column).collect { |c| c.column}
NoMethodError: private method `select' called for NetAppFiler:Class
```
I'm not worried about this one call as much as determining how to address all the other models/callers this effects.  If we convert these models to use ActiveModel do you know if that will address most of the issues?

jrafanie commented:
@gmcculloug I don't know about the details of ActiveModel.  @matthewd said in irc that it's not really covering the use cases that are added by ActsAsArModel so the short time would be to get ActsAsArModel to work with rails 3/4 finders and new methods.
